File: product-service/src/main/java/com/product/service/utils/validator/RequestValidator.java
package com.product.service.utils.validator;

import com.product.service.dto.tag.TagRequest;
import com.product.service.exception.exceptions.tag.exist.TagNameAlreadyExistException;
import com.product.service.exception.exceptions.tag.invalid.InvalidTagNameException;
import com.product.service.exception.exceptions.tag.nullable.TagNameIsNullException;
import com.product.service.exception.exceptions.tag.nullable.TagRequestIsNullException;
import com.product.service.utils.request.jdbc.accelerator.JdbcAccelerator;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class RequestValidator {
    private final JdbcAccelerator accelerator;

    public void validate(TagRequest tagRequest) {
        if (tagRequest == null) throw TagRequestIsNullException.create();
        if (tagRequest.getTagName() == null) throw TagNameIsNullException.create();
        if (tagRequest.getTagName().isBlank()) throw InvalidTagNameException.create();
    }

    public void createValidation(TagRequest tagRequest) {
        if (tagRequest == null) throw TagRequestIsNullException.create();
        if (tagRequest.getTagName() == null) throw TagNameIsNullException.create();
        if (tagRequest.getTagName().isBlank()) throw InvalidTagNameException.create();
        this.validateTagNameDuplication(tagRequest.getTagName());
    }

    public void validateTagNameDuplication(String tagName) {
        if (accelerator.isTagNameExist(tagName)) throw TagNameAlreadyExistException.create(tagName);
    }
}

File: notification-service/src/main/kotlin/ua/notification/service/utils/validator/RequestValidator.kt
package ua.notification.service.utils.validator

import org.springframework.stereotype.Component
import ua.notification.service.dto.DirectNotificationRequest
import ua.notification.service.dto.NotificationRequest
import ua.notification.service.entity.additional.NotifyType
import ua.notification.service.entity.additional.notification.NotificationPrincipal
import ua.notification.service.exception.request.disabled.UserDissembledNotificationTypeException
import ua.notification.service.exception.request.exceeded.MessageLengthExceededException
import ua.notification.service.exception.request.found.UserNotFoundException
import ua.notification.service.exception.request.invalid.InvalidMessageException
import ua.notification.service.exception.request.invalid.InvalidUserIdException
import ua.notification.service.utils.request.accelerator.JdbcAccelerator

@Component
class RequestValidator(
    private val accelerator: JdbcAccelerator
) {

    fun forceValidate(notificationRequest: NotificationRequest, type: NotifyType) {
        this.messageValidate(type, notificationRequest.message)
    }

    fun forceValidate(type: NotifyType, directNotificationRequest: DirectNotificationRequest) {
        this.messageValidate(type, directNotificationRequest.message)
        this.forceUserValidate(directNotificationRequest.userId)
        this.forceUserSettingsAccessValidate(directNotificationRequest.userId!!, type)
    }

    private fun messageValidate(type: NotifyType, message: String?) {
        if (message.isNullOrBlank()) throw InvalidMessageException("Message must be not null or blank")
        if ((type == NotifyType.ALL || type == NotifyType.SMS) && message.length >= 160)
            throw MessageLengthExceededException.create(message.length)
    }

    private fun forceUserValidate(id: Long?) {
        if (id == null) throw InvalidUserIdException("User id can't be null")
        if (accelerator.existUserById(id)) return
        throw UserNotFoundException("User not found by id: $id")
    }

    private fun forceUserSettingsAccessValidate(id: Long, type: NotifyType) {
        val principal: NotificationPrincipal = accelerator.getNotificationPrincipalByUserId(id)
            .orElseThrow{ UserNotFoundException.create(id) }
        if (!principal.emailNotification && type == NotifyType.EMAIL)
            throw UserDissembledNotificationTypeException.create(id, type)
        if (!principal.smsNotification && type == NotifyType.SMS)
            throw UserDissembledNotificationTypeException.create(id, type)
    }
}